### PR TITLE
Add support for conversion of snippets to code messages in matrix

### LINKED ISSUE #37: Slack Snippets are silently ignored
See https://get.slack.help/hc/en-us/articles/204145658-Create-a-snippet

### PR DESCRIPTION
**Pull request by Cadair**
Signed-off-by: Stuart Mumford <stuart@cadair.com>

closes #37

### CODE REVIEW

File: lib/BridgedRoom.js [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 var Promise = require('bluebird');
 
+var url = require('url');
```
Comment (Half-Shot) [inline lib/BridgedRoom.js]: ``const`` please.

File: lib/BridgedRoom.js [hunk 2]
```diff
@@ -264,39 +265,80 @@ BridgedRoom.prototype._handleSlackMessage = function(message, ghost) {
...
-                    {
-                        // Yes, we hardcode jpeg. Slack always use em.
-                        title: `${file.name}_thumb.jpeg`,
-                        mimetype: "image/jpeg",
-                    },
-                    thumb_uri,
-                    this._slack_bot_token
-                );
+            if (file.mode === "snippet") {
+                var options = url.parse(file.url_private);
```
Comment (Half-Shot) [inline lib/BridgedRoom.js]: Somewhere in the code I did this with ``rp`` and it was a lot less painful than directly using ``https``. You could save yourself a lot of code doing it :)

Comment (Half-Shot) [reply lib/BridgedRoom.js]: So https://github.com/matrix-org/matrix-appservice-slack/blob/develop/lib/SlackGhost.js#L182 I download a file with ``rp`` which you should be able to steal for your purposes?

**Comment by Cadair** [comment]
@Half-Shot this appears to work.

File: lib/BridgedRoom.js [hunk 3]
```diff
@@ -264,39 +265,81 @@ BridgedRoom.prototype._handleSlackMessage = function(message, ghost) {
...
-                    this._slack_bot_token
-                );
+            if (file.mode === "snippet") {
+                return rp({
+                    uri: file.url_private,
+                    headers: {
+                        Authorization: `Bearer ${this._slack_bot_token}`,
+                    }
+                }).then(function (htmlString) {
+                    var code = '```';
```
Comment (Half-Shot) [inline lib/BridgedRoom.js]: let

File: lib/BridgedRoom.js [hunk 4]
```diff
@@ -264,39 +265,81 @@ BridgedRoom.prototype._handleSlackMessage = function(message, ghost) {
...
-                    thumb_uri,
-                    this._slack_bot_token
-                );
+            if (file.mode === "snippet") {
+                return rp({
+                    uri: file.url_private,
+                    headers: {
+                        Authorization: `Bearer ${this._slack_bot_token}`,
+                    }
+                }).then(function (htmlString) {
```
Comment (Half-Shot) [inline lib/BridgedRoom.js]: ``(htmlString) =>  {`` please

File: lib/BridgedRoom.js [hunk 5]
```diff
@@ -264,39 +265,81 @@ BridgedRoom.prototype._handleSlackMessage = function(message, ghost) {
...
+                    var code = '```';
+                    code += '\n';
+                    code += htmlString;
+                    code += '\n';
+                    code += '```';
+                    if (file.filetype) {
+                        var html_code = '<pre><code class="language-' + file.filetype + '">';
```
Comment (Half-Shot) [inline lib/BridgedRoom.js]: define ``let htmlCode`` above, and then set it in these. Also please don't use snake case.

File: lib/BridgedRoom.js [hunk 6]
```diff
@@ -264,39 +265,81 @@ BridgedRoom.prototype._handleSlackMessage = function(message, ghost) {
...
+                    return ghost.sendMessage(roomID, content);
+                }).then(() => {
+                    // TODO: Currently Matrix lacks a way to upload a "captioned image",
+                    //   so we just send a separate `m.image` and `m.text` message
+                    // See https://github.com/matrix-org/matrix-doc/issues/906
+                    if (message.text) {
+                        return ghost.sendText(roomID, message.text);
+                    }
+                }).catch(function (err) {
+                    log.error("Unable to download snippet");
```
Comment (Half-Shot) [inline lib/BridgedRoom.js]: Log the error contents, please :)

**Pull request merged**